modules/gdscript/gdscript_compiler.cpp

```diff
@@ -2186,8 +2186,47 @@ Error GDScriptCompiler::_parse_class_level(GDScript *p_script, const GDScriptPar
 	p_script->tool = parser->is_tool();
 	p_script->name = p_class->identifier ? p_class->identifier->name : "";
 
+	if (p_script->name != "") {
+		if (ClassDB::class_exists(p_script->name) && ClassDB::is_class_exposed(p_script->name)) {
+			_set_error("The class '" + p_script->name + "' shadows a native class", p_class);
+			return ERR_ALREADY_EXISTS;
+		}
+	}
+
 	Ref<GDScriptNativeClass> native;
 
+#define CHECK_SCRIPT_MEMBER_INHERITANCE(name, member_type) GDScript *c = p_script; \
```

Comment:
Any reason you do this check on the compiler? I think it can be done on the analyzer itself (in particular in the `resolve_class_interface` method as it's already looping through members and checking stuff, inheritance is already known at that point). It has better usability since the compiler only runs when you save the script so it won't report errors as you're typing and the compiler still reports a single error while the analyzer can report all at once.

Reply:
Also, probably can be a function too.